devito/function.py

```diff
@@ -355,7 +356,12 @@ def argument_values(self, alias=None, **kwargs):
             else:
                 # We've been provided a pure-data replacement (array)
                 values[key] = new
-                # TODO: Re-derive defaults from shape of array
+                if len(new.shape) != len(self.indices):
+                    raise ValueError("Array shape %s does not match" % (new.shape, ) +
+                                     "dimensions %s" % (self.indices, ))
+                else:
+                    for i, s, o in zip(self.indices, new.shape, self.staggered):
```

Comment:
Why is this for loop different than that at line 367 ? here we also take into account ``shape`` and ``staggered``, while below only the indices. I don't know why; but in any case, at least a comment is necessary

Reply:
also, if we go through this else, it feels like we're doing the argument evaluation twice? (one here, one at line 367)

Reply:
>Why is this for loop different than that at line 367 ? 

`**kwargs` only contains the `Function` object, not the indices, so it doesn't update the same. I however should add the check to only update unset indices. Let me check if I can just remove 367